tests/integration/upgrade_tests/test_rollback.py

```diff
@@ -20,12 +72,39 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
 
 @pytest.mark.group(1)
 @pytest.mark.abort_on_fail
-async def test_upgdade(ops_test: OpsTest) -> None:
+async def test_upgrade_and_rollback(
+    ops_test: OpsTest, local_charm: pathlib.Path, faulty_upgrade_charm: pathlib.Path
+) -> None:
     """Tests that upgrade can be ran successfully."""
-    # TODO: instead of using new_charm - use the one deployed on charmhub - cannot do this until
-    # upgrades "land" on charmhub
-    new_charm = await ops_test.build_charm(".")
-    await ops_test.model.applications[MONGOS_APP_NAME].refresh(path=new_charm)
+    mongos_application = ops_test.model.applications[MONGOS_APP_NAME]
+    await mongos_application.refresh(path=faulty_upgrade_charm)
+    logger.info("Wait for upgrade to fail")
+    for attempt in tenacity.Retrying(
+        reraise=True,
+        stop=tenacity.stop_after_delay(UPGRADE_TIMEOUT),
+        wait=tenacity.wait_fixed(10),
+    ):
+        with attempt:
+            assert "Upgrade incompatible" in get_juju_status(
+                ops_test.model.name, MONGOS_APP_NAME
+            ), "Not indicating charm incompatible"
+
+    logger.info("Re-refresh the charm")
+    await mongos_application.refresh(path=local_charm)
+
+    # sleep to ensure that active status from before re-refresh does not affect below check
+    time.sleep(15)
+    await ops_test.model.block_until(
+        lambda: all(
+            unit.workload_status == "active" for unit in mongos_application.units
+        )
+        and all(unit.agent_status == "idle" for unit in mongos_application.units)
+    )
+
+    logger.info("Wait for the charm to be rolled back")
     await ops_test.model.wait_for_idle(
-        apps=[MONGOS_APP_NAME], status="active", timeout=1000, idle_period=120
+        apps=[MONGOS_APP_NAME],
+        status="active",
+        timeout=1000,
+        idle_period=30,
     )
```

Comment:
Mind adding a check to verify that mongos can still work after being rolled back? something like a read or write op is enough

Reason : we don't want to put too much faith in juju status, we want to ensure that the underying daemon still functions

Reply:
Sure